chia/full_node/mempool_manager.py

```diff
@@ -82,6 +84,69 @@ def validate_clvm_and_signature(
     return None, bytes(result), new_cache_entries
 
 
+def compute_spend_cost(
+    puzzle_reveal: SerializedProgram, solution: SerializedProgram, maximum_cost: uint64
+) -> Optional[uint64]:
+    try:
+        cost, _ = puzzle_reveal.run_with_cost(maximum_cost, solution)
+        return uint64(cost)
+    except Exception:
+        return None
+
+
+def check_item_for_dedup(
+    item: MempoolItem, dedup_coin_spends: Dict[bytes32, Tuple[bytes32, Optional[uint64]]]
+) -> Optional[Tuple[List[CoinSpend], uint64, Dict[bytes32, Tuple[bytes32, Optional[uint64]]]]]:
+    consider_this_item = True
+    cost_saving = uint64(0)
+    spends_to_dedup: List[CoinSpend] = []
+    new_dedups: Dict[bytes32, Tuple[bytes32, Optional[uint64]]] = {}
+    # See if this item has coin spends that are eligible for deduplication
+    assert item.npc_result.conds is not None
+    eligible_coin_ids = [bytes32(s.coin_id) for s in item.npc_result.conds.spends if s.eligible_for_dedup]
+    for coin_id in eligible_coin_ids:
+        # Pick the first coin spend with coin_id
+        # npc_result.conds.spends should be consistent with
+        # spend_bundle.coin_spends at this point, guaranteeing the presence of
+        # this coin spend in this bundle as a result
+        coin_spend_in_bundle = next(cs for cs in item.spend_bundle.coin_spends if cs.coin.name() == coin_id)
```

Comment:
presumably there's exactly one coin whose coin ID is `coin_id`. Instead of performing a linear search here, how about `eligible_coin_ids` just contain a pointer to the `CoinSpend` object instead?